### ISSUE TITLE
Undesirable grid element focus ring that cannot be disabled.

### ISSUE THREAD

**Issue by jflamy-dev** [vaadin/flow-components#5407]
### Description of the bug

Clicking in a grid in SINGLE selection mode to open a dialog.  When exiting the dialog, the cell that was clicked to open the dialog shows a focus ring.  This focus ring is undesirable in my user interface flow and is confusing --the user does not understand why focus is shown on this item.

This is a new v24 behavior: Vaadin v14 and v23 there is **no** focus ring on the selected cell.
Adding a cell focus listener indicates that the focus event is from the client, and not from the server itself.

See below for a sequence that reproduces the issue.

### Expected behavior

The documentation for Java Usage for v24 Grid says that focus ring is normally triggered by keyboard navigation.  There should be a way to disable this focus, or an easy way to clear it programmatically from inside the cell focus listener itself.

### Minimal reproducible example

Initial grid, with no selection.  The grid is is in single selection mode, with a listener that opens a dialog.
![image](https://github.com/vaadin/flow/assets/53801001/4d01a77a-ac9b-4868-9589-7dd48ae5b2cb)

Clicked on the M2 cell of the second row.  A modal dialog opens.  Notice there is no highlight on the cell that was clicked.
![image](https://github.com/vaadin/flow/assets/53801001/08b258c5-acbf-4cee-aec0-3d7427b87053)

Immediately exit the dialog by typing ESC. The cell clicked to select the row has focus.  This focus is undesirable and confusing to the user.
![image](https://github.com/vaadin/flow/assets/53801001/561b2335-8bd9-4c4a-a3cd-020ae89e9a4d)

### Versions

Vaadin: 24.1.4
Flow: 24.1.5
Java: JetBrains s.r.o. 17.0.6
OS: amd64 Windows 11 10.0
Browser: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36 Edg/116.0.1938.62
Live reload: Java unavailable (HotswapAgent): Front end unavailable

**Comment by knoobie** [comment]
Closing a dialog BY KEYBOARD (Escape is on the keyboard) resets the focus to the element that have opened the dialog - inclusive a visual appearance. This is a **must have** requirement for accessibility.

**Comment by web-padawan** [comment]
> Immediately exit the dialog by typing ESC. The cell clicked to select the row has focus.

This is likely a result of accessibility improvements implemented in Vaadin 24. I can confirm it's intended behavior.

**Comment by web-padawan** [comment]
> I think it might be more expected to show the focus outline on the whole row. But is that possible at the moment?

We only show outline on the whole row when switching to the "row focus" mode when navigating with keyboard.

**Comment by jflamy-dev** [comment]
If there was a way to get row focus I could live with it.  Alternately, how can I blur/move focus to the row in my focus listener?

**Comment by rolfsmeds** [comment]
We did consider implementing a way to switch the Grid permanently to row-focus mode, but abandoned the idea due to [certain complications](https://github.com/vaadin/web-components/issues/1915#issuecomment-376822852).

Although you can hide the cell focus ring with CSS, it's generally not a good idea to hide it altogether, as that will make it very difficult to navigate (and scroll) the Grid by keyboard.

Here's a somewhat hacky solution that might get you the desired behavior, however:

1. Use a combination of `PartNameGenerator` and `CellFocusEvent` (and one line of css) to highlight the entire row containing the currently focused cell;
2. Hide the cell focus ring by default...
3. ...But bring it back when the user uses the arrow keys on the Grid.

```java
private Person focusedItem = null;

public class GridView extends VerticalLayout {

    public GridView() {
    
        /* omitted grid initialization code */

        // Applies a custom part name to cells of the focused row:
        grid.setPartNameGenerator(item->{
            if (item.equals(focusedItem)) {
                return "focused-row-cell";
            } else {
                return null;
            }
        });

        // Tricks pulled when cell focus changes:
        grid.addCellFocusListener(cellFocusEvent->{
            Person oldFocusedItem = focusedItem;
            // Updates the focusedItem used by the PartNameGenerator:
            focusedItem = cellFocusEvent.getItem().orElse(null); 
            // Refreshes the old and new rows to re-run the PartNameGenerator:
            DataProvider<Person,?> dp = grid.getDataProvider();
            if (oldFocusedItem != null) dp.refreshItem(oldFocusedItem);
            if (focusedItem != null) dp.refreshItem(focusedItem);
        });

        // Applies "kbd" classname to Grid whenever arrow keys are pressed while focused:
        final List<Key> arrowKeys = Arrays.asList(Key.ARROW_DOWN, Key.ARROW_LEFT, Key.ARROW_RIGHT, Key.ARROW_UP);
        for (Key key : arrowKeys) {
            UI.getCurrent().addShortcutListener(e->{
                grid.addClassName("kbd");
            }, key).listenOn(grid);
        }

        // Removes that classname whenever the Grid is focused or blurred:
        grid.addBlurListener(e->{
            grid.removeClassName("kbd");
        });
        grid.addFocusListener(e->{
            grid.removeClassName("kbd");
        });
    }
}
```

And the CSS:

```css
/* Hides cell focus ring unless "kbd" classname is applied: */
vaadin-grid:not(.kbd)::part(focused-cell)::before {
    display: none;
}

/* Applies a subtle background color to the focused row's cells: */
vaadin-grid::part(focused-row-cell) {
    background: var(--lumo-contrast-5pct);
}
```

There may very well be caveats and gotchas with this code, and there may very well be better ways to achieve this, but it seems to work decently enough in my testing so far.

**Comment by rolfsmeds** [comment]
Come to think of it, there's an alternate solution that replaces the cellFocusListener + partNameGeneartor part of the above solution above with CSS, but it doesn't work in Firefox yet (but will later this year):

This needs to go into `components/vaadin-grid.css` in your theme folder.
```css
[part~="focused-cell"],
td:has(~ [part~="focused-cell"]),
[part~="focused-cell"] ~ td {
    background: var(--lumo-contrast-5pct);
}```

**Comment by jflamy** [comment]
There has to be a way to get rid of the focus ring programmatically.   In my use case, user clicks on row 5.  Edits the item, which causes the grid to be sorted again.  The item on row 5 is now on row 1.   Having a cell selected on row 5 when users use the mouse (and used the mouse to close the dialog -- not ever using the keyboard navigation) is completely unnatural and confusing.

I witnessed it first hand and telling people "just ignore the highlight" is not satisfactory.
Having read the "simplified styling" article, I would have thought that putting
```
vaadin-grid::part(focused-row-cell) {
   display: none;
}
```
in the styles.css file under the themes/myapp folder would have worked, but that was naïve.   Where is it supposed to go, and is that the newfangled style where the grid is the parent style and the part a child ?

**Comment by rolfsmeds** [comment]
Frankly, I don't see how the solution containing a few lines of CSS is a big problem for you. "Messing" with stylesheets is no more hacky than "messing" with Java code to achieve some kind of workaround.

But anyway, if I'm understanding you correctly, it sounds like you don't want _any_ focus in the Grid when the dialog closes. How about blurring the Grid on focus whenever the dialog has been open?

- Have a boolean `preventGridFocus`, default to false
- Set it true when the dialog opens
- Have a FocusListener on the Grid, which, if `preventGridFocus` is true, blurs the Grid.

There will be a brief flash of a focus ring on the cell, but after that the Grid will lose focus and the focus ring is gone.

**Comment by jflamy-dev** [comment]
Grids that are sorted based on a dialog end up with the wrong item focused - the item clicked is irrelevant to the item that needs to be selected upon exit.  It should be trivial to reselect the correct item, and keyboard navigation should be reset to that new reference, not to the old one.

I did try blurring the grid, but the focus ring stubbornly remained.  I could have done the blurring wrong.  But "move keyboard focus to match select" would be really nice.

**Comment by rolfsmeds** [comment]
The problem is that with a lazy-loading, infinite-scrolling Grid, the Grid doesn't actually know which row each item is on. This is also why there are no methods e.g. for scrolling to a specific item (see https://github.com/vaadin/web-components/issues/1951).

After all, the row might even be outside of the currently visible viewport, which would require the Grid to automatically scroll up or down until the item comes into view.

We're looking into some way to get around that limitation, but I can assure you it is not, in fact, trivial.

Super-crude code here, but it seems to work for me:
```
        g.addItemClickListener(e->{
            dd.open();
            isDialogOpen = true;
        });

        g.addFocusListener(event -> {
            if (isDialogOpen) {
                g.blur();
                isDialogOpen = false;
            }
        });
```

**Comment by jflamy-dev** [comment]
Thanks.  For the "For Flow you don't need JavaScript and CSS" dogma, the above recipe is fine.
